python/ray/tune/tests/test_result_grid.py

```diff
@@ -36,8 +37,41 @@ def f(config):
     assert result.checkpoint is None
 
 
+def test_best_result():
+    def f(config):
+        for _ in range(2):
+            tune.report(x=config["x"])
+
+    analysis = tune.run(f, config={"x": tune.grid_search([1, 2])})
+    result_grid = ResultGrid(analysis)
+    best_result = result_grid.get_best_result(metric="x", mode="max")
+    assert best_result.config["x"] == 2
+    assert best_result.metrics["x"] == 2
+
+
+def test_best_result_no_report():
+    def f(config):
+        pass
+
+    analysis = tune.run(f, config={"x": tune.grid_search([1, 2])})
+    result_grid = ResultGrid(analysis)
+    with pytest.raises(RuntimeError):
+        result_grid.get_best_result(metric="x", mode="max")
+
+
+def test_no_metric_mode():
+    def f(config):
+        tune.report(x=1)
+
+    analysis = tune.run(f)
+    result_grid = ResultGrid(analysis)
+    with pytest.raises(ValueError):
+        result_grid.get_best_result()
```

Comment:
Should we separate them? Since all should raise an error?

Reply:
Ah yeah good point...updated!